devito/at_controller.py

```diff
@@ -0,0 +1,225 @@
+
+        self.report_dir = at_report_dir if at_report_dir is not None else default_at_dir
+        self.final_report_path = path.join(self.report_dir, final_report_name)
+
+        if not path.isdir(self.report_dir):  # Creates report dir if does not exist
+            mkdir(self.report_dir)
+
+    @staticmethod
+    def get_at_block_size(f_name, time_order, spc_border, shape, block_dims, at_report_dir=None):
+        """
+        Gets block size from auto tuning report
+        :param f_name: str - function name. Used for naming
+        :param time_order: int - time order of kernel
+        :param spc_border: int - space border of kernel
+        :param shape: list - shape of the data buffer
+        :param block_dims: list - indicating which dims are blocked.
+        :param at_report_dir: string - indicating path to auto tuning report directory.
+                              If not set default at report directory is used
+        :raises ValueError: if auto tuning report is not found
+        """
+        global final_report_name, default_at_dir, model_descr_template
+
+        report_dir = at_report_dir if at_report_dir is not None else default_at_dir
+        final_report_path = path.join(report_dir, final_report_name)
+
+        if not path.isfile(final_report_path):
+            raise ValueError("Auto tuning report at %s not found" % final_report_path)
+
+        # Cache blocking dimensions
+        cb_dims = str([True if block else False for block in block_dims]).replace(" ", '')
+        # model description string
+        model_descr_str = model_descr_template % (f_name, time_order, spc_border,
+                                                  str(shape).replace(" ", ''), cb_dims)
+
+        with open(final_report_path, 'r') as f:
+            for line in f.readlines():
+
+                if model_descr_str in line:
+                    blocks_str = line.split(' ')[5]
+                    block_split = blocks_str[1:len(blocks_str) - 2].split(',')
+
+                    return [int(block) if block != "None" else None for block in block_split]
+
+        return None  # returns none if no matching block size was found
+
+    @staticmethod
+    def get_optimal_block_size(shape, load_c):
+        """
+        Gets optimal block size based on architecture
+        :param shape: list - shape of the data buffer
+        :param load_c: int - load count
+        :return: optimal block size
+        """
+        cache_s = int(cpuinfo.get_cpu_info()['l2_cache_size'].split(' ')[0])  # cache size
+        core_c = cpuinfo.get_cpu_info()['count']  # number of cores
+
+        # assuming no prefetching, square block will give the most cache reuse.
+        # We then take the cache/core divide by the size of the inner most dimension in which we do not block.
+        #  This gives us the X*Y block space, of which we take the square root to get the size of our blocks.
+        # ((C size / cores) / (4 * length inner most * kernel loads)
```

Comment:
This description should really go in the docstring above to make it available to everyone without looking at the code.